app/controllers/concerns/link_permissions_checks.rb

```diff
@@ -0,0 +1,56 @@
+require "active_support/concern"
+
+module LinkPermissionsChecks
+  extend ActiveSupport::Concern
+
+  # Use callbacks to share common setup or constraints between actions.
+  def set_link(link_id)
+    if link_id.match? /\D+/
+      @link = Link.find_by(custom_url: link_id)
+    else
+      @link = Link.find_by(id: link_id)
+    end
+  end
+
+  def require_link
+    if @link.nil?
+      raise ActionController::RoutingError.new('Not Found')
```

Comment:
I added a pattern of separating "setting" from "enforcing" just to make it possible to write code where it's okay for the param to not resolve (is this an antipattern in rails?)